File: src/com/leetcode/Medium/SpiralMatrixII.java
package com.leetcode.Medium;

// 59. Spiral Matrix II
public class SpiralMatrixII {
    public int[][] generateMatrix(int n) {
        int[][] matrix = new int[n][n];
        if (n == 0)
            return matrix;


        int i, k = 0, l = 0, c = matrix[0].length, r = matrix.length;
        int j = 1;
        while (k < r && l < c || j < n * n) {
            for (i = l; i < c; i++) {
                matrix[k][i] = j++;
            }
            k++;

            for (i = k; i < r; i++) {
                matrix[i][c - 1] = j++;
            }
            c--;

            if (k < r) {
                for (i = c - 1; i >= l; i--) {
                    matrix[r - 1][i] = j++;
                }
                r--;
            }

            if (l < c) {
                for (i = r - 1; i >= k; i--) {
                    matrix[i][l] = j++;
                }
                l++;
            }
        }
        return matrix;
    }
}

File: JavaScript/Medium/SpiralMatrixII.js
/**
 * 59. Spiral Matrix II
 * @param {number} n
 * @return {number[][]}
 */
var generateMatrix = function (n) {
  const matrix = new Array(n).fill().map(() => new Array(n).fill(0));

  let top = 0;
  let bottom = n - 1;
  let left = 0;
  let right = n - 1;

  let current = 1;
  while (top <= bottom && left <= right) {
    // Traverse the top row from left to right
    for (let i = left; i <= right; i++) {
      matrix[top][i] = current++;
    }
    top++;

    // Traverse the right column from top to bottom
    for (let i = top; i <= bottom; i++) {
      matrix[i][right] = current++;
    }
    right--;

    // Traverse the bottom row from right to left
    for (let i = right; i >= left; i--) {
      matrix[bottom][i] = current++;
    }
    bottom--;

    // Traverse the left column from bottom to top
    for (let i = bottom; i >= top; i--) {
      matrix[i][left] = current++;
    }
    left++;
  }

  return matrix;
};
